core/src/org/sbml/jsbml/Unit.java

```diff
@@ -1041,182 +1041,101 @@ public static boolean isValidUnit(Model model, String unit) {
-            newExponent = unit1.getExponent() + unit2.getExponent();
-            if (newExponent == 0d) {
-              newMultiplier = 1d;
-            } else if (unit1.getMultiplier() != unit2.getMultiplier()) {
-              m1 = unit1.getMultiplier();
-              m2 = unit2.getMultiplier();
-              e1 = unit1.getExponent();
-              e2 = unit2.getExponent();
-              newMultiplier = Math.pow(Math.pow(m1, e1) * Math.pow(m2, e2), 1d / newExponent);
-            }
-          }
-        }
+    double newExponent;
+    double newMultiplier;
 
-        // Adapt unit kind if necessary
-        if (k1 == Kind.METER) {
-          unit1.setKind(Kind.METRE);
-        } else if (k1 == Kind.LITER) {
-          unit1.setKind(Kind.LITRE);
-        }
+    Kind kind1 = unit1.getKind();
+    Kind kind2 = unit2.getKind();
 
-      } else {
-        /*
-         * If the unit has become dimentionless, mark it accordingly
-         * However, we must keep scale and multiplier even in dimensionless
-         * units because these could become important in later operations.
-         */
-        newMultiplier = Math.pow(m1, e1) * Math.pow(m2, e2); // 1d;
-        newScale = s1 * e1 + s2 * e2; // 0d;
-        newExponent = 1d;
-        unit1.setKind(Kind.DIMENSIONLESS);
-      }
+    // Adapt unit kinds if necessary
+    if (kind1 == Kind.METER) {
+      unit1.setKind(Kind.METRE);
+    } else if (kind1 == Kind.LITER) {
+      unit1.setKind(Kind.LITRE);
+    }
+    if (kind2 == Kind.METER) {
+      unit2.setKind(Kind.METRE);
+    } else if (kind2 == Kind.LITER) {
+      unit2.setKind(Kind.LITRE);
+    }
 
-      unit1.setMultiplier(newMultiplier);
-      unit1.setScale((int) newScale);
-      unit1.setExponent(newExponent);
 
-    } else if (k1.equals(Kind.INVALID) || k2.equals(Kind.INVALID)) {
+    // If one of the units are invalid then an invalid unit is returned with the parameters
+    // of
+    if(unit1.getKind().equals(Kind.INVALID) || unit2.getKind().equals(Kind.INVALID)) {
 
       // The resulting units is always invalid
       unit1.setKind(Kind.INVALID);
-      
-      // TODO - just set all other attributes to their default for invalid units ?
-      
-      if (!k2.equals(Kind.INVALID)) {
+
+      if (!unit2.getKind().equals(Kind.INVALID)) {
         if (unit2.isSetOffset) {
           unit1.setOffset(unit2.getOffset());
         }
         unit1.setMultiplier(unit2.getMultiplier());
         unit1.setScale(unit2.getScale());
         unit1.setExponent(unit2.getExponent());
       }
-    } else {
-      throw new IllegalArgumentException(MessageFormat.format(
-        "Cannot merge units with different kind properties {0} and {1}. Units can only be merged if both have the same kind attribute or if one of them is dimensionless.",
-        k1, k2));
+
+      return;
+    }
+
+    // If units are not invalid but not of same kind don't merge
+    if(!unit1.getKind().equals(unit2.getKind())) {
+      return;
+    }
```

Comment:
An aditional comment on that. Not sure if you are planning to work on validation as well but you might want to make sure that unit validation is not impacted by changes in behavior. It would be good if not already done that @eikept has the offline validation checks setup so that he can see if unit validation is affected and if validation need to be ajusted after the changes.

Reply:
So I have now looked at the ValidationTests and found that the tests in `checkConsistencyOfflineTests` all pass. My question would be if I should also test it with `OfflineValidatorTests`?  If I understand correctly I would need to set up a test folder structure with a folder for each error code for that. Is that correct?

Reply:
Since it has now been a month I wanted to ask again if somebody could answer my questions from above? Thank you in advance.

Reply:
Unfortunately, I don't know about the testing strategy. @niko-rodrigue and @draeger could probably answer this.